### PR TITLE
ENH: Make Xorbits Data iterable

### LINKED ISSUE #72: ENH: Make it more friendly when iterating `DataFrame.columns`
I'm thinking to add a property `own_data` to indicate whether xorbits object holds the data directly, for those objects created from pandas or numpy, iterating and printing them doesn't need execution, just iterating or printing the pandas dataframe or numpy ndarray.

Here list the methods that can skip execution for "own-data" entity:
- items
- iterrows
- itertuples
- `__str__`
- `__repr__`

As Mars hasn't implemented `__iter__`, so does xorbits.

### PR DESCRIPTION
**Pull request by aresnow1**
## What do these changes do?

- Make Xorbits Data iterable.
- Skip execution for those who owns data.

## Related issue number

Fixes #72 

## Check code requirements

- [x] tests added / passed (if needed)
- [x] Ensure all linting tests pass

### CODE REVIEW

File: python/xorbits/core/data.py [hunk 1]
```diff
@@ -177,17 +175,60 @@ def __setattr__(self, key, value):
         else:
             self.data.__setattr__(key, value)
 
+    def own_data(self):
```
Comment (qianduoduo0904) [inline python/xorbits/core/data.py:178]: IMO this kind of method should not be exposed to public, at least for now, we can use `_own_data` instead.

File: python/xorbits/web/tests/test_request_web.py [hunk 2]
```diff
@@ -22,7 +22,7 @@
 
 @pytest.fixture
 def init_with_web():
-    sess = init_test(n_cpu=2, web=True)
```
Comment (qianduoduo0904) [inline python/xorbits/web/tests/test_request_web.py:25]: Why remove `n_cpu=2`?

Comment (aresnow1) [reply python/xorbits/web/tests/test_request_web.py:25]: This case is to check if web works, seems doesn't need multi-processing cluster.

**Review by qianduoduo0904 — APPROVED** [review]
LGTM

File: python/xorbits/core/data.py [hunk 3]
```diff
@@ -177,17 +175,60 @@ def __setattr__(self, key, value):
         else:
             self.data.__setattr__(key, value)
 
+    def _own_data(self):
+        from .adapter import own_data
+
+        return own_data(self.data._mars_entity)
+
+    def __iter__(self):
+        # Mars entity hasn't implemented __iter__, however `iter(mars_entityt)`
```
Comment (UranusSeven) [inline python/xorbits/core/data.py:184]:
```suggestion
        # Mars entity hasn't implemented __iter__, however `iter(mars_entity)`
```

File: python/xorbits/pandas/mars_adapters/__init__.py [hunk 4]
```diff
@@ -101,3 +102,20 @@ def _install():
         MarsDataFrameGroupBy.transform, "transform", DataType.dataframe_groupby
     )
     register_data_members(DataType.dataframe_groupby, dataframe_groupby_udfs)
+
+    # install iteration functions
+    # DataFrame.itertuples, DataFrame.itertuples
```
Comment (UranusSeven) [inline python/xorbits/pandas/mars_adapters/__init__.py:107]: The second `itertuples` should be `iterrows`?

File: python/xorbits/pandas/mars_adapters/tests/test_execution.py [hunk 5]
```diff
@@ -70,3 +72,31 @@ def test_from_mars_execution_condition(setup, dummy_df, dummy_str_series):
         # TODO: series falling back methods not implemented yet.
         # assert not need_to_execute(dummy_str_series.to_pickle(os.path.join(tempdir, "series.pkl")))
         # assert not need_to_execute(dummy_str_series.to_json())
+
+
+def test_own_data(setup, dummy_df, dummy_str_series):
```
Comment (UranusSeven) [inline python/xorbits/pandas/mars_adapters/tests/test_execution.py:77]: Better add some tests to check if the docstring works.

File: python/xorbits/pandas/mars_adapters/__init__.py [hunk 6]
```diff
@@ -101,3 +102,20 @@ def _install():
...
+    dataframe_iteration_functions = dict()
+    dataframe_iteration_functions["itertuples"] = wrap_iteration_functions(
+        MarsDataFrame.itertuples, "itertuples", True
+    )
+    dataframe_iteration_functions["iterrows"] = wrap_iteration_functions(
```
Comment (UranusSeven) [inline python/xorbits/pandas/mars_adapters/__init__.py:112]: Do we also need to wrap DataFrame.items?

https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.items.html#pandas.DataFrame.items

File: python/xorbits/pandas/mars_adapters/core.py [hunk 7]
```diff
@@ -203,3 +200,27 @@ def wrapped(*args, **kwargs):
...
+    @functools.wraps(member_method)
+    def wrapped(self: MarsEntity, *args, **kwargs):
+        if own_data(self):
+            # if own data, return iteration on data directly
+            return getattr(self.op.data, member_name)(*to_mars(args), **to_mars(kwargs))
+        else:
+            return from_mars(member_method(self, *to_mars(args), **to_mars(kwargs)))
+
+    if attach_docstring:
+        return attach_cls_member_docstring(wrapped, member_name)
```
Comment (UranusSeven) [inline python/xorbits/pandas/mars_adapters/core.py:223]: The docstring attaching seemed to be failed. I checked the docstring of `iterrows` and the docstring was emtpy.

File: python/xorbits/pandas/mars_adapters/tests/test_execution.py (hunk 5 above)
Comment (aresnow1) [reply python/xorbits/pandas/mars_adapters/tests/test_execution.py:77]: Added.

File: python/xorbits/pandas/mars_adapters/core.py (hunk 7 above)
Comment (aresnow1) [reply python/xorbits/pandas/mars_adapters/core.py]: Fixed.

File: python/xorbits/pandas/mars_adapters/__init__.py (hunk 6 above)
Comment (aresnow1) [reply python/xorbits/pandas/mars_adapters/__init__.py:112]: `DataFrame.items` Iterates over the DataFrame columns, it won't trigger execution in Mars, I prefer to keep this behavior as one column data would be huge, return a Xorbits Series is better.

**Review by UranusSeven — APPROVED** [review]
LGTM.

**Review by qianduoduo0904 — APPROVED** [review]
LGTM

**Pull request merged**